docs/proposals/20191017-control-plane.md

```diff
@@ -0,0 +1,494 @@
+1. In service of user story 5, the control plane provider should also manage etcd membership via kubeadm as part of scaling up and down.
+1. The control plane provider should provide indicators of health to meet user story 6 and 10. This should include at least the state of etcd and information about which replicas are currently healthy or not. For the default implementation, health attributes based on artifacts kubeadm installs on the cluster may also be of interest to cluster operators.
+1. The control plane provider should be able to upgrade a control plane’s version of Kubernetes as well as updating the underlying machine image on cloud installations.
+
+### Implementation Details/Notes/Constraints
+
+#### New API Types
+
+ControlPlane:
+
+```go
+package v1alpha3
+
+import (
+    corev1  "k8s.io/api/core/v1"
+    metav1  "k8s.io/apimachinery/pkg/apis/meta/v1"
+    cabpkv1 "sigs.k8s.io/cluster-api-bootstrap-provider-kubeadm/api/v1alpha2"
+)
+
+// ControlPlaneSpec defines the desired state of ControlPlane
+type ControlPlaneSpec struct {
+    // Number of desired machines. Defaults to 1.
+    // This is a pointer to distinguish between explicit zero and not specified.
+    Replicas int32 `json:"replicas"`
+
+    // Version defines the desired Kubernetes version.
+    Version string `json:"version"`
+
+    // InfrastructureTemplate is a required reference to a custom resource
+    // offered by an infrastructure provider.
+    InfrastructureTemplate corev1.ObjectReference `json:"infrastructureTemplate"`
+
+    // KubeadmConfigSpec is a KubeadmConfigSpec
+    // to use for initializing and joining machines to the control plane
+    KubeadmConfigSpec cabpkv1.KubeadmConfigSpec `json:"kubeadmConfigSpec"`
+}
+
+// ControlPlaneStatus defines the observed state of ControlPlane
+type ControlPlaneStatus struct {
+  // Selector is the label selector in string format to avoid introspection
+  // by clients, and is used to provide the CRD-based integration for the
+  // scale subresource and additional integrations for things like kubectl
+  // describe.. The string will be in the same format as the query-param syntax.
+  // More info about label selectors: http://kubernetes.io/docs/user-guide/labels#label-selectors
+  // +optional
+  Selector string `json:"selector,omitempty"`
+
+  // Total number of non-terminated machines targeted by this control plane
+  // (their labels match the selector).
+  // +optional
+  Replicas int32 `json:"replicas,omitempty"`
+
+  // Total number of non-terminated machines targeted by this control plane
+  // that have the desired template spec.
+  // +optional
+  UpdatedReplicas int32 `json:"updatedReplicas,omitempty"`
+
+  // Total number of fully running machines, including the number of active
```

Comment:
I'm confused about the etcd portion of this comment - this potentially sounds like we're saying the value here would be 6 for a 3-machine control plane (3 machines + 3 etcd cluster members).